components/layout/construct.rs

```diff
@@ -674,21 +669,23 @@ impl<'a, ConcreteThreadSafeLayoutNode: ThreadSafeLayoutNode>
     /// Pushes fragments appropriate for the content of the given node onto the given list.
     fn create_fragments_for_node_text_content(&self,
                                               fragments: &mut IntermediateInlineFragments,
-                                              node: &ConcreteThreadSafeLayoutNode,
-                                              style: &Arc<ServoComputedValues>) {
+                                              node: &ConcreteThreadSafeLayoutNode) {
         // Fast path: If there is no text content, return immediately.
         let text_content = node.text_content();
         if text_content.is_empty() {
             return
         }
 
-        let mut style = (*style).clone();
-        match node.get_pseudo_element_type() {
-            PseudoElementType::Before(_) |
-            PseudoElementType::After(_) => {}
-            _ => properties::modify_style_for_text(&mut style)
-        }
+        let node_is_input_or_text_area =
+           node.type_id() == Some(LayoutNodeType::Element(LayoutElementType::HTMLInputElement)) ||
+           node.type_id() == Some(LayoutNodeType::Element(LayoutElementType::HTMLTextAreaElement));
 
+        let context = self.style_context();
+        let mut style = node.style(context);
+        if node_is_input_or_text_area {
```

Comment:
Fwiw, per spec text input and textarea are replaced elements.  How they're implemented is a separate issue, of course.